saas_demo/models/saas_operator.py

```diff
@@ -78,23 +80,32 @@ def update_addons_path(self):
     @api.multi
     def restart_odoo(self):
         if self.is_local():
+            test = self.env['ir.config_parameter'].get_param('test_saas_demo')
+            if test:
+                # no need to restart odoo folder in test mode
+                return
             service.server.restart()
 
     @api.multi
     def _update_repos(self):
         self.ensure_one()
         if self.type != 'local':
             return
-        return self.demo_id.repo_ids._local_update_repo()
-
-    @api.multi
-    def update_ad_paths(self, path):
-        if self.is_local():
-            if path not in ad_paths:
-                ad_paths.append(path)
-
-    @api.multi
-    def remove_ad_paths(self, path):
-        if self.is_local():
-            if path in ad_paths:
-                ad_paths.remove(path)
+        has_updates = False
+        for repo in self.demo_id.repo_ids:
+            updated = self.server_update_repo(repo.url, repo.url_escaped, repo.branch, repo.commit)
+            if updated:
+                has_updates = True
+        return has_updates
+
+    @staticmethod
+    def server_update_repo(url, url_escaped, branch, commit):
+        repos_root = repos_dir()
```

Comment:
Add a code description with explanation what the method does (git pull ?)

Reply:
This is base method from os.py which creates repos directory

Reply:
Or are you asking about server_update_repo?

Reply:
about  local_server_update_repo